sklearn/preprocessing/data.py

```diff
@@ -54,41 +58,25 @@
 )
 
 
-def _mean_and_std(X, axis=0, with_mean=True, with_std=True):
-    """Compute mean and std deviation for centering, scaling.
-
-    Zero valued std components are reset to 1.0 to avoid NaNs when scaling.
-    """
-    X = np.asarray(X)
-    Xr = np.rollaxis(X, axis)
-
-    if with_mean:
-        mean_ = Xr.mean(axis=0)
-    else:
-        mean_ = None
-
-    if with_std:
-        std_ = Xr.std(axis=0)
-        std_ = _handle_zeros_in_scale(std_)
-    else:
-        std_ = None
-
-    return mean_, std_
-
-
-def _handle_zeros_in_scale(scale):
+def _handle_zeros_in_scale(scale, copy=True):
     ''' Makes sure that whenever scale is zero, we handle it correctly.
 
     This happens in most scalers when we have constant features.'''
 
     # if we are fitting on 1D arrays, scale might be a scalar
     if np.isscalar(scale):
-        if scale == 0:
+        if scale == .0:
             scale = 1.
+        return scale
     elif isinstance(scale, np.ndarray):
-        scale[scale == 0.0] = 1.0
-        scale[~np.isfinite(scale)] = 1.0
```

Comment:
Infinite values from the input are filter out by `check_array`. In case those are produced by our computation, we should not silently remove them anyway.